File: python/src/ctrldesign/mpc/mpc_problem.py
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from typing import Optional, Protocol

import numpy as np

from ..ode_modelling.casadi_helper import casadi_vars_to_str
from ..ode_modelling.ode_model import OdeModel

FixedValueArg = float | str


@dataclass
class FixedValue:
    numeric_value: Optional[float]
    parameter_name: Optional[str]

    def is_numeric(self) -> bool:
        return self.numeric_value is not None

    def get_numeric_value(self) -> float:
        return self.numeric_value

    def get_parameter_name(self) -> str:
        return self.parameter_name

    @staticmethod
    def Parameter(name: str):
        return FixedValue(numeric_value=None, parameter_name=name)

    @staticmethod
    def Numeric(value: float):
        return FixedValue(numeric_value=value, parameter_name=None)


@dataclass
class FreeInput:
    signal: str
    min: float
    max: float
    init: float


@dataclass
class FixedInput:
    signal: str
    value: FixedValue


@dataclass
class Constraint:
    signal: str
    min: float
    max: float
    feedthrough: bool


@dataclass
class QuadraticCost:
    signal: str
    ref: FixedValue
    weight_stage: float
    weight_end: float
    scaled_weights: bool
    feedthrough: bool


@dataclass
class QuadraticDiffQuotCost:
    signal: str
    quot: tuple[float]
    ts_order: int
    weight: float
    scaled_weight: bool
    feedthrough: bool
    prev_value: Optional[FixedValue]


@dataclass
class LinearCost:
    signal: str
    weight_stage: bool
    weight_end: bool
    scaled_weights: bool
    feedthrough: bool


@dataclass
class Parameter:
    signal: str
    is_vector: bool
    scaled_like: str


class SignalProp(Protocol):
    signal: str


def _warn(msg: str):
    print(f"Waring: {msg}")


def get_signals(v: Iterable[SignalProp]) -> list[str]:
    return [x.signal for x in v]


def find_signal_idx(v: Iterable[SignalProp], signal: str) -> Optional[int]:
    for i, el in enumerate(v):
        if el.signal == signal:
            return i

    return None


class MpcProblem:
    def __init__(
        self,
        model: OdeModel,
    ):
        self.model = model

        self.u_names = casadi_vars_to_str(self.model.inputs)
        self.x_names = casadi_vars_to_str(self.model.states)
        self.y_names = self.model.y_names

        self.parameters: list[Parameter] = []

        self.free_inputs: list[FreeInput] = []
        self.fixed_inputs: list[str] = []
        self.constraints: list[Constraint] = []
        self.costs_quadratic: list[QuadraticCost] = []
        self.costs_quadratic_diffquot: list[QuadraticDiffQuotCost] = []
        self.costs_linear: list[LinearCost] = []

    @property
    def nu_free(self) -> int:
        return len(self.free_inputs)

    def add_free_input(
        self, name: str, min: float, max: float, init: Optional[float] = None
    ):
        if name not in self.u_names:
            raise ValueError(f"'{name}' is not an input of the model")

        if (idx := find_signal_idx(self.free_inputs, name)) is None:
            self.free_inputs.append(
                FreeInput(name, min, max, np.nan if init is None else init)
            )

        else:
            _warn(
                f"'{name}' was already defined as a free input. "
                "Updating its parameters, but the previous index of the input remains valid."
            )
            self.free_inputs[idx](
                FreeInput(name, min, max, np.nan if init is None else init)
            )

    def add_fixed_input(self, name: str, value: FixedValueArg):
        if name not in self.u_names:
            raise ValueError(f"'{name}' is not an input of the model")

        if find_signal_idx(self.free_inputs, name) is not None:
            _warn(f"'{name}' is already a free input. The fixed value will be ignored.")

        if (idx := find_signal_idx(self.fixed_inputs, name)) is None:
            self.fixed_inputs.append(value)
        else:
            self.fixed_inputs[idx] = value

    def add_fixed_inputs(self, fixed_inputs: dict[str, FixedValueArg]):
        for k, v in fixed_inputs.items():
            self.add_fixed_input(k, v)

    def add_constraint(
        self, name: str, min: float, max: float, feedthrough: Optional[bool] = None
    ):
        if name in self.u_names:
            raise ValueError(
                f"'{name}' is an input. Constraints on inputs must be specified with add_free_input"
            )

        if min > max:
            raise ValueError("invalid constraint: max is smaller than min")

        feedthrough = self._check_signal(name, feedthrough)

        for i, c in enumerate(self.constraints):
            if (name == c.signal) and (feedthrough == c.feedthrough):
                _warn(
                    f"For '{name}' a constraint with the same feedthrough value was already defined and gets overwritten now."
                )

                self.constraints[i] = Constraint(name, min, max, feedthrough)
                self.dirty = True
                return

        self.constraints.append(Constraint(name, min, max, feedthrough))

    def add_quadratic_cost(
        self,
        name: str,
        ref: float | str,
        weight_stage: float = 0.0,
        weight_end: float = 0.0,
        feedthrough: Optional[bool] = None,
        scaled_weights: bool = False,
    ):
        feedthrough = self._check_signal(name, feedthrough)

        ref = self._check_fixed_value(ref, name)
        cost = QuadraticCost(
            name, ref, weight_stage, weight_end, scaled_weights, feedthrough
        )

        for i, c in enumerate(self.costs_quadratic):
            if (name == c.signal) and (feedthrough == c.feedthrough):
                _warn(
                    f"For '{name}' a quadratic cost with the same feedthrough value was already defined and gets overwritten now."
                )

                self.costs_quadratic[i] = cost
                return

        self.costs_quadratic.append(cost)

    def add_quadratic_diffquot_cost(
        self,
        name: str,
        diff_quot: Sequence[float],
        ts_order: Optional[int] = None,
        weight: float = 0.0,
        feedthrough: Optional[bool] = None,
        scaled_weight: bool = False,
        prev_value: Optional[FixedValueArg] = None,
    ):
        feedthrough = self._check_signal(name, feedthrough)

        diff_quot = tuple(diff_quot)

        if not np.isclose(sum(diff_quot), 0):
            _warn(f"difference quotient factors for '{name}' don't sum up to 0.")

        if ts_order is None:
            ts_order = 0.0

        if prev_value is not None:
            prev_value = self._check_fixed_value(prev_value, name)

        cost = QuadraticDiffQuotCost(
            name, diff_quot, ts_order, weight, scaled_weight, feedthrough, prev_value
        )

        for i, c in enumerate(self.costs_quadratic_diffquot):
            if (name == c.signal) and (feedthrough == c.feedthrough):
                _warn(
                    f"For '{name}' a quadratic diff quot cost with the same feedthrough value was already defined and gets overwritten now."
                )

                self.costs_quadratic_diffquot[i] = cost
                return

        self.costs_quadratic_diffquot.append(cost)

    def add_linear_cost(
        self,
        name: str,
        weight_stage: float,
        weight_end: float,
        feedthrough: Optional[bool] = None,
        scaled_weights: bool = False,
    ):
        feedthrough = self._check_signal(name, feedthrough)

        cost = LinearCost(name, weight_stage, weight_end, scaled_weights, feedthrough)

        for i, c in enumerate(self.costs_quadratic):
            if (name == c.signal) and (feedthrough == c.feedthrough):
                self.costs_linear[i] = cost
                return

        self.costs_linear.append(cost)

    def _check_fixed_value(self, value: FixedValueArg, ref_signal: str) -> FixedValue:
        if not isinstance(value, str):
            return FixedValue.Numeric(value)

        name = value.strip()

        if not name.startswith("#"):
            raise ValueError("Fixed value must be numerical or start with '#'")

        name = name[1:]

        is_vector = name.endswith("[]")

        if is_vector:
            name = name[:-2]

        if name in self.x_names:
            raise ValueError(
                f"'{name}' is already a state and cannot be used as parameter name"
            )

        if name in self.u_names:
            raise ValueError(
                f"'{name}' is already an input and cannot be used as parameter name"
            )

        if name in self.y_names:
            raise ValueError(
                f"'{name}' is already an output and cannot be used as parameter name"
            )

        if name in {"x0", "u", "lam_g"}:
            raise ValueError(f"'{name}' is a reserved name")

        if find_signal_idx(self.parameters, name) is None:
            self.parameters.append(Parameter(name, is_vector, ref_signal))
        else:
            raise ValueError(f"'{name}' is already used as a parameter name")

        return FixedValue.Parameter(name)

    def _check_signal(self, name: str, feedthrough: Optional[bool]) -> bool:
        if name in self.u_names:
            if feedthrough is not None:
                raise ValueError(
                    "The parameter 'feedthrough' can only be used with outputs, "
                    f"but '{name}' is an input."
                )

            feedthrough = True

        elif name in self.x_names:
            if feedthrough is not None:
                raise ValueError(
                    "The parameter 'feedthrough' can only be used with outputs, "
                    f"but '{name}' is a state."
                )

            feedthrough = False

        elif name in self.y_names:
            feedthrough = False if feedthrough is None else feedthrough

        else:
            raise ValueError(f"'{name}' is not a signal of the model.")

        return feedthrough

    def get_all_constraints(self) -> dict[str, tuple[float, float]]:
        constraints = {v.signal: (v.min, v.max) for v in self.constraints}
        input_constraints = {v.signal: (v.min, v.max) for v in self.free_inputs}

        constraints.update(input_constraints)

        return constraints

    def check(self) -> bool:
        defined_inputs = set(
            get_signals(self.free_inputs) + get_signals(self.fixed_inputs)
        )

        model_inputs = set(self.u_names)

        missing_inputs = model_inputs - defined_inputs

        if len(missing_inputs) > 0:
            print(
                "the following model inputs are neither defined as free nor fixed inputs: "
                f"{', '.join(missing_inputs)}"
            )
            return False

        return True

File: matlab/mpc/MpcProblem.m
classdef MpcProblem

    properties
        model

        u_names
        x_names
        y_names

        free_inputs = struct('signal', {}, 'min', {}, 'max', {}, 'init', {})
        fixed_inputs = struct('signal', {}, 'value', {})
        constraints = struct('signal', {}, 'min', {}, 'max', {}, 'feedthrough', {})
        costs_quadratic = struct('signal', {}, 'ref', {}, ...
            'weight_stage', {}, 'weight_end', {}, 'scaled_weights', {}, 'feedthrough', {})
        costs_quadratic_diffquot = struct('signal', {}, 'quot', {}, 'ts_order', {}, ...
            'weight', {}, 'scaled_weight', {}, 'feedthrough', {}, 'prev_value', {})
        costs_linear = struct('signal', {}, ...
            'weight_stage', {}, 'weight_end', {}, 'scaled_weights', {}, 'feedthrough', {})

        parameters = struct('signal', {}, 'is_vector', {}, 'scaled_like', {});
    end
    
    properties (Dependent)
        nu_free
    end

    methods
        function obj = MpcProblem(model)
            obj.model = model;
            obj.u_names = casadi_vars_to_str(obj.model.inputs);
            obj.x_names = casadi_vars_to_str(obj.model.states);
            obj.y_names = obj.model.y_names;            

        end

        function value = get.nu_free(obj)
            value = length(obj.free_inputs);
        end

        function obj = add_free_input(obj, name, min, max, init)
            if (nargin < 5), init = nan; end

            [~, id] = ismember(name, obj.u_names);

            if id == 0
                error('"%s" is not an input of the model', name);
            end

            idx = find_signal(obj.free_inputs, name);

            if idx == 0
                obj.free_inputs(end + 1) = struct('signal', name, 'min', min, 'max', max, 'init', init);
            else
                warning("'%s' was already defined as free input. " + ...
                        "Updating its parameters, but the previous index of the input remains valid.");
            
                obj.free_inputs(idx) = struct('signal', name, 'min', min, 'max', max, 'init', init);
            end
        end

        function obj = add_fixed_input(obj, name, value)
            if ~ismember(name, obj.u_names)
                error("'%s' is nont an input of the model", name);
            end

            if find_signal(obj.free_inputs, name) ~= -1
                warning("'%s' is already a free input. The fixed value will be ignored.");
            end

            idx = find_signal(obj.fixed_inputs, name);

            if idx == 0
                obj.fixed_inputs(end + 1) = struct('signal', signal, 'value', value);
            else
                obj.fixed_inputs(idx) = struct('signal', signal, 'value', value);
            end
        end

        function obj = add_fixed_inputs(obj, fixed_inputs)
            names = fieldnames(fixed_inputs);

            for i = 1:length(names)
                obj = obj.add_fixed_input(names{i}, fixed_inputs.(names{i}));
            end
        end

        function obj = add_constraint(obj, name, min, max, varargin)
            ip = inputParser();
            ip.KeepUnmatched = false;
            ip.addParameter("feedthrough", [], @islogical);
            ip.parse(varargin{:});
            feedthrough = ip.Results.feedthrough;

            if ismember(name, obj.u_names)
                error("'%s' is an input. Constraints on inputs must be specified with add_free_input", name);
            end

            if min > max
                error("invlaid constraint: max is smaller than min");
            end

            feedthrough = obj.check_signal_(name, feedthrough);
            
            for i = 1:length(obj.constraints)
                if strcmp(obj.constraints(i).signal, name) && (obj.constraints(i).feedthrough == feedthrough)
                    warning("For '%s' a constraint with the same feedthrough value was already defined and gets overwritten now.", ...
                        name);
        
                    obj.constraints(i) = struct('signal', name, 'min', min, 'max', max, 'feedthrough', feedthrough);
                    return;
                end
            end

            obj.constraints(end + 1) = struct('signal', name, 'min', min, 'max', max, 'feedthrough', feedthrough);
        end

        function obj = add_quadratic_cost(obj, name, ref, varargin)
            ip = inputParser();
            ip.KeepUnmatched = false;
            ip.addParameter("weight_stage", 0, @isnumeric);
            ip.addParameter("weight_end", 0, @isnumeric);
            ip.addParameter("feedthrough", [], @islogical);
            ip.addParameter("scaled_weights", false, @islogical);
            ip.parse(varargin{:});
            weight_stage = ip.Results.weight_stage;
            weight_end = ip.Results.weight_end;
            feedthrough = ip.Results.feedthrough;
            scaled_weights = ip.Results.scaled_weights;

            feedthrough = obj.check_signal_(name, feedthrough);

            [ref, param] = obj.check_fixed_value_(ref, name);
    
            if ~isempty(param)
                obj.parameters(end + 1) = param;
            end
            

            cost = struct('signal', name, 'ref', ref, ...
                'weight_stage', weight_stage, 'weight_end', weight_end, ...
                'scaled_weights', scaled_weights, 'feedthrough', feedthrough);

            for i = 1:length(obj.costs_quadratic)
                if strcmp(obj.costs_quadratic(i).signal, name) && (obj.costs_quadratic(i).feedthrough == feedthrough)
                    warning("For '%s' a quadratic cost with the same feedthrough value was already defined and gets overwritten now.", ...
                        name);
               
                    obj.costs_quadratic(i) = cost;
                    return;
                end
            end
            obj.costs_quadratic(end + 1) = cost;

        end

        function obj = add_quadratic_diffquot_cost(obj, name, diff_quot, varargin)
            ip = inputParser();
            ip.KeepUnmatched = false;
            ip.addParameter("ts_order", 0, @isnumeric);
            ip.addParameter("weight", 0, @isnumeric);
            ip.addParameter("feedthrough", [], @islogical);
            ip.addParameter("scaled_weight", false, @islogical);
            ip.addParameter("prev_value", []);
            ip.parse(varargin{:});
            ts_order = ip.Results.ts_order;
            weight = ip.Results.weight;
            feedthrough = ip.Results.feedthrough;
            scaled_weight = ip.Results.scaled_weight;
            prev_value = ip.Results.prev_value;

            feedthrough = obj.check_signal_(name, feedthrough);

            if abs(sum(diff_quot)) > 1e-12
                warning("different quotient factors for '%s' don't sum up to 0", name);
            end

            if ~isempty(prev_value)
                [prev_value, param] = obj.check_fixed_value_(prev_value, name);

                if ~isempty(param)
                    obj.parameters(end + 1) = param;
                end
            end

            cost = struct('signal', name, ...
                'quot', diff_quot, 'ts_order', ts_order, ...
                'weight', weight, 'scaled_weight', scaled_weight, ...
                'feedthrough', feedthrough, 'prev_value', prev_value);

            for i = 1:length(obj.costs_quadratic_diffquot)
                if strcmp(obj.costs_quadratic_diffquot(i).signal, name) ...
                        && (obj.costs_quadratic_diffquot(i).feedthrough == feedthrough)
                    warning("For '%s' a quadratic diff quot cost with the same feedthrough value was already defined and gets overwritten now.", ...
                        name);

                    obj.costs_quadratic_diffquot(i) = cost;
                    return;
                end
            end

            obj.costs_quadratic_diffquot(end + 1) = cost;

        end

        function obj = add_linear_cost(obj, name, varargin)
            ip = inputParser();
            ip.KeepUnmatched = false;
            ip.addParameter("weight_stage", 0, @isnumeric);
            ip.addParameter("weight_end", 0, @isnumeric);
            ip.addParameter("feedthrough", [], @islogical);
            ip.addParameter("scaled_weights", false, @islogical);
            ip.parse(varargin{:});
            weight_stage = ip.Results.weight_stage;
            weight_end = ip.Results.weight_end;
            feedthrough = ip.Results.feedthrough;
            scaled_weights = ip.Results.scaled_weights;

            feedthrough = obj.check_signal_(name, feedthrough);

            cost = struct('signal', name, ...
                'weight_stage', weight_stage, 'weight_end', weight_end, ...
                'scaled_weights', scaled_weights, 'feedthrough', feedthrough);

            for i = 1:length(obj.costs_linear)
                if strcmp(obj.costs_linear(i).signal, name) && (obj.costs_linear(i).feedthrough == feedthrough)
                    warning("For '%s' a quadratic cost with the same feedthrough value was already defined and gets overwritten now.", ...
                        name);

                    obj.costs_linear(i) = cost;
                    return;
                end
            end

            obj.costs_linear(end + 1) = cost;

        end

        function cons = get_all_constraints(obj)
            cons = struct();

            for i = 1:length(obj.constraints)
                c = obj.constraints(i);
                cons.(c.signal).min = c.min;
                cons.(c.signal).max = c.max;
            end

            for i = 1:length(obj.free_inputs)
                u = obj.free_inputs(i);
                cons.(u.signal).min = u.min;
                cons.(u.signal).max = u.max;
            end
        end
    end

    methods(Hidden)
        function [value, param] = check_fixed_value_(obj, value, ref_signal)
            if isnumeric(value)
                value = FixedValue.Numeric(value);
                param = [];
                return;
            end

            name = strtrim(value);

            if isempty(value) || (value(1) ~= '#')
                error("Fixed value must be numerical or start with '#'");
            end

            name = name(2:end);

            is_vector = (length(value) > 2) && strcmp(value(end-1:end), '[]');

            if is_vector
                name = name(1:end-2);
            end

            if ismember(name, obj.x_names)
                error("parameter name '%s' is already a state", name);
            elseif ismember(name, obj.u_names)
                error("parameter name '%s' is already an input", name);
            elseif ismember(name, obj.y_names)
                error("parameter name '%s' is already an output", name);
            elseif ismember(name, {'x0', 'u', 'lam_g'})
                error("parameter name '%s' is a reserved name", pname);
            end

            idx = find_signal(obj.parameters, name);

            if (idx == 0)
                param = struct('signal', name, 'is_vector', is_vector, 'scaled_like', ref_signal);
            else
                error("'%s' is already used as a parameter name");
            end
            
            value = FixedValue.Parameter(name);
        end

        function feedthrough = check_signal_(obj, name, feedthrough)
            if ismember(name,  obj.u_names)
                if ~isempty(feedthrough)
                    error("The parameter 'feedthrough' can only used with outputs, but" ...
                          + " '%s' is an input.", name);
                end
                feedthrough = true;

            elseif ismember(name, obj.x_names)
                if ~isempty(feedthrough)
                    error("The parameter 'feedthrough' can only used with or outputs, but" ...
                          + " '%s' is a state.", name);
                end
                feedthrough = false;

            elseif ismember(name, obj.model.y_names)
                if isempty(feedthrough)
                    feedthrough = false;
                end

            else
                error("'%s' is not a signal of the model.", name);
            end
        end
    end
end
